File: src/c_interface/libnegf_c_interface.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <complex.h>
#include "libnegf_c_interface.h"

// -------------------------------------------------------------------
// // to compile this code linking to the shared library
// gcc main.c -o my_app -L. -lLibNEGFCInterface -Wl,-rpath,.
// -------------------------------------------------------------------

// Helper function to generate a random double between 0.0 and 1.0
double rand_double() {
    return (double)rand() / (double)RAND_MAX;
}

int main() {
    // Seed the random number generator
    srand(time(NULL));

    printf("--- Generating Block-Tridiagonal Complex CSC Matrix ---\n");

    // 1. Matrix parameters
    int32_t n_blocks = 10;
    int32_t block_size = 64;
    int32_t m = n_blocks * block_size; // 640
    int32_t n = n_blocks * block_size; // 640
    
    // Calculate exact number of non-zeros (nnz)
    // A block tridiagonal matrix has:
    // - 1 main diagonal of blocks (N blocks)
    // - 1 super-diagonal of blocks (N-1 blocks)
    // - 1 sub-diagonal of blocks (N-1 blocks)
    // Total blocks = 3*N - 2. Elements per block = b^2.
    int32_t total_blocks = 3 * n_blocks - 2;
    int32_t nnz = total_blocks * (block_size * block_size); // 114,688

    printf("Matrix Dimensions: %d x %d\n", m, n);
    printf("Total Non-Zero Elements: %d\n", nnz);

    // 2. Allocate memory for CSC arrays
    int32_t* colptr_in = (int32_t*)malloc((n + 1) * sizeof(int32_t));
    int32_t* rowval_in = (int32_t*)malloc(nnz * sizeof(int32_t));
    double complex* nzval_in = (double complex*)malloc(nnz * sizeof(double complex));

    if (!colptr_in || !rowval_in || !nzval_in) {
        printf("Memory allocation failed!\n");
        return 1;
    }

    // 3. Populate the CSC arrays
    int32_t nz_count = 0;
    
    for (int32_t j = 0; j < n; j++) {
        colptr_in[j] = nz_count;
        
        int32_t block_col = j / block_size; // Which block column are we in?

        // Determine which block rows contain data for this block column
        int32_t block_row_start = (block_col > 0) ? block_col - 1 : 0;
        int32_t block_row_end = (block_col < n_blocks - 1) ? block_col + 1 : n_blocks - 1;

        // Iterate through the valid block rows
        for (int32_t K = block_row_start; K <= block_row_end; K++) {
            // Iterate through the actual rows within this block
            for (int32_t r = 0; r < block_size; r++) {
                int32_t i = K * block_size + r; // Actual row index (0-based)
                
                rowval_in[nz_count] = i;
                
                // Assign a random complex value (a + b*I)
                nzval_in[nz_count] = rand_double() + rand_double() * I;
                
                nz_count++;
            }
        }
    }
    colptr_in[n] = nz_count; // Cap off the colptr array

    // 4. Construct Block Sizes Vector
    int32_t* block_sizes_in = (int32_t*)malloc(n_blocks * sizeof(int32_t));
    for (int i = 0; i < n_blocks; i++) {
        block_sizes_in[i] = block_size; // Populate [64, 64, ..., 64]
    }

    // 5. Construct the Dictionary inputs
    int32_t n_dict = 2;
    const char* dict_keys_in[] = {"max_iter", "tolerance"};
    int32_t dict_vals_in[] = {1000, 1};

    // 6. Boolean 
    int32_t is_hermitian = 1;

    printf("\n--- Calling Julia Library ---\n\n");

    // 7. Call the Julia function
    CCscMatrix result = run_bndiag_wrapper(
        m, n, nnz, colptr_in, rowval_in, nzval_in,
        n_blocks, block_sizes_in,
        n_dict, dict_keys_in, dict_vals_in,
        is_hermitian
    );

    printf("\n--- Back in C: Reading Results ---\n");
    printf("Returned Matrix Size: %d x %d\n", result.m, result.n);
    printf("Returned Non-zeros: %d\n", result.nnz);

    // Print just the first few values to verify
    if (result.nnz > 0) {
        printf("First 3 Complex Values Returned:\n");
        for (int i = 0; i < 3 && i < result.nnz; i++) {
            printf("  [%d]: %f + %fi\n", i, creal(result.nzval[i]), cimag(result.nzval[i]));
        }
    }

    // 8. CRITICAL: Cleanup Memory
    // Free the arrays we dynamically allocated for the inputs
    free(colptr_in);
    free(rowval_in);
    free(nzval_in);
    free(block_sizes_in);

    // Free the arrays Julia allocated and passed back to us
    free(result.colptr);
    free(result.rowval);
    free(result.nzval);

    printf("\nMemory freed successfully. Exiting.\n");
    return 0;
}

File: src/c_interface/libnegf_c_interface.jl
# ------------------------------------------------------------------------------
# instructions to compile and create the shared library file

# NOTE : it seems that the main directory needs to be renamed
# to LibNEGF.jl

# # install JuliaC
# # 1. from the terminal
# LibNEGF.jl% julia
# # 2. open Julia's REPL
# julia> ]
# # 3. install JuliaC
# (@v1.12) pkg> app add JuliaC

# # then, go to the root directory of this project (which has been
# # renamed at this point as LibNEGF.jl) and run from the terminal:
# LibNEGF.jl% ./compile.sh cpu

# FIXME : is the following commented block still relevant?
# # above, the flag --export-ai generate a JSON file, which in turn can be used to
# # generate the header for the shared library .. I don't think this is really needed
# # at the moment, right? Because we have manually created our own interface header
# using JuliaLibWrapping
# # Translates the JSON spec into LibNEGFCInterface.h
# generate_c_header("LibNEGFCInterface.json", "LibNEGFCInterface.h")
# ------------------------------------------------------------------------------

# wrapper for RGF
function bndiag_of_inv_rgf_local_wrapper(MspIN::SparseArrays.SparseMatrixCSC{FieldType,Int}, blockSizes::Vector{Int},
    ndiag::Dict{String,Int}, isHermitian::Bool)::SparseArrays.SparseMatrixCSC{FieldType,Int}

    println(Core.stdout, "Running sequential RGF from its C interface function")

    # # convert the sparse input matrix to BlockMatrix type first
    MbmIN::BlockMatrix = bm_convert(MspIN, blockSizes, ndiag, isHermitian)
    MbmOUT::BlockMatrix = bm_copy(MbmIN)
    # allocate auxiliary data
    auxData::AuxDataRGF = allocate_aux_data_RGF(MbmIN)
    # call RGF
    bndiag_of_inv_rgf_local!(MbmOUT, MbmIN, auxData, TimingData(), CountingData())
    # create and return the output
    MspOUT::SparseArrays.SparseMatrixCSC{FieldType,Int} = bm_convert(MbmOUT)

    return MspIN
end

# C-COMPATIBLE RETURN STRUCT
# Note the change to Ptr{FieldType} for the non-zero values
struct CCscMatrix
    m::Cint
    n::Cint
    nnz::Cint
    colptr::Ptr{Cint}
    rowval::Ptr{Cint}
    nzval::Ptr{FieldType}
end

# THE C-CALLABLE WRAPPER
Base.@ccallable function run_bndiag_wrapper(
    # CSC Matrix Inputs
    m::Cint, n::Cint, nnz_in::Cint,
    colptr_in::Ptr{Cint}, rowval_in::Ptr{Cint}, nzval_in::Ptr{FieldType},

    # Vector Inputs
    n_blocks::Cint, block_sizes_in::Ptr{Cint},

    # Dictionary Inputs
    n_dict::Cint, dict_keys_in::Ptr{Cstring}, dict_vals_in::Ptr{Cint},

    # Boolean Input
    is_hermitian::Cint
)::CCscMatrix

    # --- A. RECONSTRUCT JULIA OBJECTS FROM C POINTERS ---
    colptr_j = unsafe_wrap(Array, colptr_in, n + 1; own=false) .+ 1
    rowval_j = unsafe_wrap(Array, rowval_in, nnz_in; own=false) .+ 1
    nzval_j = unsafe_wrap(Array, nzval_in, nnz_in; own=false) # Now an Array of FieldType
    MspIN = SparseMatrixCSC(Int(m), Int(n), colptr_j, rowval_j, nzval_j)

    blockSizes = Int.(unsafe_wrap(Array, block_sizes_in, n_blocks; own=false))

    keys_jv = unsafe_wrap(Array, dict_keys_in, n_dict; own=false)
    vals_jv = unsafe_wrap(Array, dict_vals_in, n_dict; own=false)
    ndiag = Dict{String,Int}()
    for i in 1:n_dict
        ndiag[unsafe_string(keys_jv[i])] = Int(vals_jv[i])
    end

    isHermitian = is_hermitian != 0

    # --- B. EXECUTE THE CORE ALGORITHM ---
    out_csc::SparseArrays.SparseMatrixCSC{FieldType,Int} = bndiag_of_inv_rgf_local_wrapper(MspIN, blockSizes, ndiag, isHermitian)

    # --- C. PREPARE THE RETURN VALUE FOR C ---
    out_nnz = SparseArrays.nnz(out_csc)
    out_m, out_n = size(out_csc)

    out_colptr_ptr = Libc.malloc(sizeof(Cint) * (out_n + 1))
    out_rowval_ptr = Libc.malloc(sizeof(Cint) * out_nnz)
    # Ensure memory is sized for FieldType (which is 16 bytes: two 8-byte floats)
    out_nzval_ptr = Libc.malloc(sizeof(FieldType) * out_nnz)

    out_colptr_0based = Cint.(out_csc.colptr .- 1)
    out_rowval_0based = Cint.(out_csc.rowval .- 1)
    out_nzval_complex = FieldType.(out_csc.nzval)

    unsafe_copyto!(Ptr{Cint}(out_colptr_ptr), pointer(out_colptr_0based), out_n + 1)
    unsafe_copyto!(Ptr{Cint}(out_rowval_ptr), pointer(out_rowval_0based), out_nnz)
    unsafe_copyto!(Ptr{FieldType}(out_nzval_ptr), pointer(out_nzval_complex), out_nnz)

    return CCscMatrix(
        Cint(out_m), Cint(out_n), Cint(out_nnz),
        Ptr{Cint}(out_colptr_ptr), Ptr{Cint}(out_rowval_ptr), Ptr{FieldType}(out_nzval_ptr)
    )
end